Mathlib/NumberTheory/LSeries/Linearity.lean

```diff
@@ -0,0 +1,107 @@
+
+lemma LSeriesSummable.add {f g : ℕ → ℂ} {s : ℂ} (hf : LSeriesSummable f s)
+    (hg : LSeriesSummable g s) :
+    LSeriesSummable (f + g) s := by
+  simpa only [LSeriesSummable, ← term_add_apply] using Summable.add hf hg
+
+@[simp]
+lemma LSeries.add {f g : ℕ → ℂ} {s : ℂ} (hf : LSeriesSummable f s) (hg : LSeriesSummable g s) :
+    LSeries (f + g) s = LSeries f s + LSeries g s := by
+  simpa only [LSeries, term_add, Pi.add_apply] using tsum_add hf hg
+#align nat.arithmetic_function.l_series_add LSeries.add
+
+/-!
+### Negation
+-/
+
+lemma LSeries.term_neg (f : ℕ → ℂ) (s : ℂ) : term (-f) s = -term f s := by
+  ext ⟨- | n⟩
+  · simp only [Nat.zero_eq, term_zero, Pi.neg_apply, neg_zero]
+  · simp only [term_of_ne_zero (Nat.succ_ne_zero _), Pi.neg_apply, Nat.cast_succ, neg_div]
+
+lemma LSeries.term_neg_apply (f : ℕ → ℂ) (s : ℂ) (n : ℕ) : term (-f) s n = -term f s n := by
+  rw [term_neg, Pi.neg_apply]
+
+lemma LSeriesHasSum.neg {f : ℕ → ℂ} {s a : ℂ} (hf : LSeriesHasSum f s a) :
+    LSeriesHasSum (-f) s (-a) := by
+  simpa only [LSeriesHasSum, term_neg] using HasSum.neg hf
+
+lemma LSeriesSummable.neg {f : ℕ → ℂ} {s : ℂ} (hf : LSeriesSummable f s) :
+    LSeriesSummable (-f) s := by
+  simpa only [LSeriesSummable, term_neg] using Summable.neg hf
+
+@[simp]
+lemma LSeriesSummable.neg_iff {f : ℕ → ℂ} {s : ℂ} :
+    LSeriesSummable (-f) s ↔ LSeriesSummable f s :=
+  ⟨fun H ↦ neg_neg f ▸ H.neg, .neg⟩
+
+@[simp]
+lemma LSeries.neg (f : ℕ → ℂ) (s : ℂ) : LSeries (-f) s = -LSeries f s := by
+  simp only [LSeries, term_neg_apply, tsum_neg]
+
+/-!
+### Scalar multiplication
+-/
+
+lemma LSeries.term_smul (f : ℕ → ℂ) (c s : ℂ) : term (c • f) s = c • term f s := by
+  ext ⟨- | n⟩
+  · simp only [Nat.zero_eq, term_zero, Pi.smul_apply, smul_eq_mul, mul_zero]
+  · simp only [term_of_ne_zero (Nat.succ_ne_zero _), Pi.smul_apply, smul_eq_mul, Nat.cast_succ,
+      mul_div_assoc]
+
+lemma LSeries.term_smul_apply (f : ℕ → ℂ) (c s : ℂ) (n : ℕ) :
+    term (c • f) s n = c * term f s n := by
+  rw [term_smul, Pi.smul_apply, smul_eq_mul]
+
+lemma LSeriesHasSum.smul {f : ℕ → ℂ} (c : ℂ) {s a : ℂ} (hf : LSeriesHasSum f s a) :
+    LSeriesHasSum (c • f) s (c * a) := by
+  simpa only [LSeriesHasSum, term_smul, smul_eq_mul] using hf.const_smul c
+
+lemma LSeriesSummable.smul {f : ℕ → ℂ} (c : ℂ) {s : ℂ} (hf : LSeriesSummable f s) :
+    LSeriesSummable (c • f) s := by
+  simpa only [LSeriesSummable, term_smul, smul_eq_mul] using hf.const_smul c
+
+lemma LSeriesSummable.of_smul {f : ℕ → ℂ} {c s : ℂ} (hc : c ≠ 0) (hf : LSeriesSummable (c • f) s) :
+    LSeriesSummable f s := by
+  simpa only [ne_eq, hc, not_false_eq_true, inv_smul_smul₀] using hf.smul (c⁻¹)
+
+lemma LSeriesSummable.smul_iff {f : ℕ → ℂ} {c s : ℂ} (hc : c ≠ 0) :
+    LSeriesSummable (c • f) s ↔ LSeriesSummable f s :=
+  ⟨fun H ↦ H.of_smul hc, fun H ↦ H.smul c⟩
+
+@[simp]
+lemma LSeries.smul (f : ℕ → ℂ) (c s : ℂ) : LSeries (c • f) s = c * LSeries f s := by
```

Comment:
Is `smul_apply` perhaps a better name?

Reply:
Maybe. But then it should also be `LSeries.add_apply` and `LSeries.neg_apply`, and one could have
```lean
lemma LSeries.smul (f : ℕ → ℂ) (c : ℂ) : LSeries (c • f) = c • LSeries f := ...
lemma LSeries.neg (f : ℕ → ℂ) : LSeries (-f) = -LSeries f := ...
```
This does not work for `add`, however, because of possibly differing convergence behaviour.

On the other hand, the other statements (`LSeries.term_smul`, `LSeriesHasSum.smul`, `LSeriesSummable.smul` etc.) all include the `s` argument, so the present version is reasonably consistent (and `LSeries.term_smul_apply` adds the `n` argument, not the `s` argument...).

Reply:
But `LSeriesSummable` and `LSeriesHasSum` are `Prop`s, while `LSeries` contains data, so the situation is not quite comparable. I would have expected `LSeries.smul` to be the map `ℂ → LSeries → LSeries` that defines the scalar multiplication operation (by analogy with `Nat.add` etc).

Reply:
PS: Possibly a better comparison is that there is `HasSum.const_smul` and `Summable.const_smul`, but there is no `tsum.const_smul`; there are no lemmas in the `tsum` namespace, because the output of `tsum` is a number, it doesn't have type `tsum`, so you can't dot-notate on it.

In fact this suggests (to me) that the naming convention `LSeries.foo` is misguided for all values of `foo` (not just `add`, `smul` etc), and they should all be `LSeries_foo` or indeed `lseries_foo`.

Reply:
@Ruben-VandeVelde suggested the change from `LSeries_smul` to `LSeries.smul` (and the same for `neg`) a few days ago.
I don't have strong feelings either way, but I would like to have some kind of consensus here, so that we can move on...

Reply:
> In fact this suggests (to me) that the naming convention `LSeries.foo` is misguided for all values of `foo` (not just `add`, `smul` etc), and they should all be `LSeries_foo` or indeed `lseries_foo`.

Enabling dot notation is not the only reason to put stuff in namespaces, though. My understanding is that this also servers as an organising principle (e.g., in a currently active Zulip thread, it is dicsussed to introduce a `Discrete` namespace to serve as a home for statements on discrete Fourier transform etc., even though there won't be things of type `Discrete ...` to which they apply).